examples/real_function_space.py

```diff
@@ -115,26 +115,60 @@ def u_exact(x):
 
 # We can now assemble the matrix and vector
 
-A = dolfinx.fem.petsc.assemble_matrix_block(a)
+try:
+    A = dolfinx.fem.petsc.assemble_matrix_block(a)
+except AttributeError:
+    A = dolfinx.fem.petsc.assemble_matrix(a, kind="mpi")
 A.assemble()
-b = dolfinx.fem.petsc.assemble_vector_block(L, a, bcs=[])
+
+
+# On the main branch of DOLFINx, the `assemble_vector` function for blocked spaces has been rewritten to reflect how
+# it works for standard assembly and `nest` assembly. This means that lifting is applied manually.
+# In this case, with no Dirichlet BC, we could skip those steps.
+# However, for clarity we include them here.
+
+bcs = []
+main_assembly = False
+try:
+    b = dolfinx.fem.petsc.assemble_vector_block(L, a, bcs=bcs)
+except AttributeError:
+    main_assembly = True
+    b = dolfinx.fem.petsc.assemble_vector(L, kind="mpi")
+    bcs1 = dolfinx.fem.bcs_by_block(dolfinx.fem.extract_function_spaces(a, 1), bcs)  # type: ignore
+    dolfinx.fem.petsc.apply_lifting(b, a, bcs=bcs1)  # type: ignore
+    b.ghostUpdate(PETSc.InsertMode.ADD, PETSc.ScatterMode.REVERSE)  # type: ignore
+    bcs0 = dolfinx.fem.bcs_by_block(dolfinx.fem.extract_function_spaces(L), bcs)  # type: ignore
+    dolfinx.fem.petsc.set_bc(b, bcs0)
+    b.ghostUpdate(PETSc.InsertMode.INSERT, PETSc.ScatterMode.FORWARD)
 
 # Next, we modify the second part of the block to contain `h`
-# We start by enforcing the multiplier constraint $h$ by modifying the right hand side vector
+# We start by enforcing the multiplier constraint $h$ by modifying the right hand side vector.
+# On the main branch, this is greatly simplified
 
-if dolfinx.__version__ == "0.8.0":
-    maps = [(V.dofmap.index_map, V.dofmap.index_map_bs), (R.dofmap.index_map, R.dofmap.index_map_bs)]
-elif Version(dolfinx.__version__) >= Version("0.9.0.0"):
-    maps = [(Wi.dofmap.index_map, Wi.dofmap.index_map_bs) for Wi in W.ufl_sub_spaces()]
 
-b_local = get_local_vectors(b, maps)
-b_local[1][:] = h
-scatter_local_vectors(
-        b,
-        b_local,
-        maps,
-    )
-b.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+uh = dolfinx.fem.Function(V, name="u")
+
+if main_assembly:
```

Comment:
What does `main_assembly mean`?

Reply:
That we are using assembly for main branch